New post - Continuous Stuff with Github Actions (#306)
subvisual/subvisual.com: src/posts/2020-04-16-continuous-stuff-with-github-actions/index.md

---
id: 147
path: /posts/continuous-stuff-with-github-actions
title: "Continuous Stuff with Github Actions"
author: miguel-palhas
date: 2020-04-20
tags:
  - devops

> intro: "Last year, I took on the task of improving the continuous process over at Utrust. We weren't really happy with the amount of work that went into our releases, and I was looking for a more agile approach, where everyone from developers to the QA team could do their part with low friction."

Reviewer: You can turn this into a multiline string to prevent the wrapping in your editor and make it easier to edit. (https://yaml-multiline.info/)

```suggestion
intro: >
  Last year, I took on the task of improving the continuous process over at
  Utrust. We weren't really happy with the amount of work that went into our
  releases, and I was looking for a more agile approach, where everyone from
  developers to the QA team could do their part with low friction.
```

---

[slides]: https://slides.com/naps62/continuous-stuff
[utrust]: https://utrust.com
[github-actions]: https://github.com/features/actions
[circleci]: http://circleci.com/
[primer]: https://www.imdb.com/title/tt0390384/
[github-deployments]: https://developer.github.com/v3/repos/deployments/
[github-deployment-status]: https://developer.github.com/v3/repos/deployments/
[github-workflow-triggers]: https://help.github.com/en/actions/reference/events-that-trigger-workflows
[circleci-orbs]: https://circleci.com/orbs/
[orb-awscli]: https://circleci.com/orbs/registry/orb/circleci/aws-cli#orb-source
[actions-checkout]: https://github.com/actions/checkout/tree/v2.0.0
[twitter]: https://twitter.com/naps62

> <div class="custom-paragraph">

Reviewer: Forgot to add this suggestion to match the one in the stylesheets.

```suggestion
<div class="BlogPostParagraph">
```

<em>
This post was adapted from a talk that I gave at several local meetups over the past few months. It got enough positive

> feedback that I felt compeled to convert it to written format. Original slides are available <a
> href="https://slides.com/naps62/continuous-stuff" target="_blanks">here</a>.

Reviewer: You don't need to use literal anchor tags to get the blank target, our posts will automatically render links like this (thanks to `gatsby-remark-external-links`, added in #286).

Also just noticed a typo: compel**l**ed :grimacing: 

```suggestion
feedback that I felt compelled to convert it to written format. Original slides are available [here][slides].
```

Author: the literal anchor tag is because I can't use markdown inside html tags

---
id: 148
path: /posts/continuous-stuff-with-github-actions
title: "Continuous Stuff with Github Actions"
author: miguel-palhas
date: 2020-04-20
tags:
  - devops
intro: >
  Last year, I took on the task of improving the continuous process over at

>   Utrust. We weren't really happy with the amount of work that went into our

Reviewer 2: Isn't it `utrust` now?

Author: it's Utrust now. It was UTRUST before

Reviewer 2: I distinctly remember someone from their team correcting us and asking to use "utrust". The [website](https://utrust.com/) is inconsistent. Title says "Utrust", most sections use "utrust". 🤔

Author: according to Jorge, Utrust is the name, utrust is the logo

  developers to the QA team could do their part with low friction.
---

[slides]: https://slides.com/naps62/continuous-stuff
[utrust]: https://utrust.com
[github-actions]: https://github.com/features/actions
[circleci]: http://circleci.com/
[primer]: https://www.imdb.com/title/tt0390384/
[github-deployments]: https://developer.github.com/v3/repos/deployments/
[github-deployment-status]: https://developer.github.com/v3/repos/deployments/
[github-workflow-triggers]: https://help.github.com/en/actions/reference/events-that-trigger-workflows
[circleci-orbs]: https://circleci.com/orbs/
[orb-awscli]: https://circleci.com/orbs/registry/orb/circleci/aws-cli#orb-source
[actions-checkout]: https://github.com/actions/checkout/tree/v2.0.0
[twitter]: https://twitter.com/naps62

<div class="BlogPostParagraph">
  <em>
    This post was adapted from a talk that I gave at several local meetups over the
    past few months. It got enough positive feedback that I felt compeled to convert it to written format. Original slides
    are available <a href="https://slides.com/naps62/continuous-stuff" target="_blank">here</a>.
  </em>
</div>

> Last year, I took on the task of improving the continuous process over at [UTRUST][utrust]. We weren't really happy with

Reviewer 2: Same as previous

Author: here it is wrong :+1:

the amount of work that went into our releases, and I was looking for a more agile approach, where everyone from
developers to the QA team could do their part with low friction.

> Coincidentally, that project started right around the same time where GitHub Actions went into public beta. Some of our

Reviewer 2: where or when? Does it make sense to go with `around the same time when GitHub Actions ...`

Author: actually it makes sense without either of them. `around the same time Github Actions went`

It's when, but it's implicit that way

problems were related to shortcomings on our existing CI solution, so it felt right to see what GitHub had to offer.

I ended up pleasantly surprised. But first, let's talk about what exactly was wrong with our previous CI.

## Shortcomings of traditional CIs

I've worked with a fair amount of CIs over the years...

> ![CIs](./cis.png)

Reviewer 2: idk if it's an issue with GitHub but this image isn't displaying for me.

Author: really? to me it is showing
both on [github](https://github.com/subvisual/subvisual.com/blob/mp/ci-post/src/posts/2020-04-16-continuous-stuff-with-github-actions/index.md) as well as in the preview on netlify

Reviewer 2: Rich diff wasn't showing, it is now. Caching issue probs. 👍

The main use case for this is the now common one: to run tests for every new version of your code.  But any kind of
automated task can be triggered, really. A deploy is very common as well, or a preview build for testing.

But a lot of these automations shouldn't necessarily need a commit. They are not triggered by changes in the code.
Deploys might be triggered by some higher-level decision, or a QA team manually approving the latest version, which was
committed days ago.

### Not everything is a commit

This led to the common practice of creating "special" ways to trigger these commits. In [CircleCI][circleci], for
example, you'd do something like this:

```yaml
workflows:
  build:
    release:
      - deploy_to_staging:
          filters:
            branches:
              only:
                - staging
```

With this, commits to the `staging` branch would trigger a deploy to staging. Seems simple enough.  But now you need
a branch for every single environment. And probably quite a few `push -f` commands, or similar git sorceries to force
a commit from your normal workflow into a branch whose history is messier than the plot of [Primer][primer].

## GitHub API to the rescue

> The problem here is that we're solving things the wrong way. A deploy is not a commit. It's a deploy.

Reviewer 2: I always have this question: is it deploy or deployment? I thought the former was the verb and the latter the noun.

Author: to be honest, I don't think there's a strict definition.

we also push "commits" and not "commitments"

Reviewer 2: Ok, that's how I've been going on about it 👍

unless a Pull Request is open for them. The rationale here is that a work-in-progress branch rarely needs CI feedback,
and would only waste resources. This is often a configurable option in CI settings. But with GitHub action triggers, we
can do a bit more than that, by having that configuration in the source code itself:

```yaml
name: GitHub Action Example

on:
  push:
    branches:
      - master
    tags:
      - '*'
  pull_request:
    types: [opened, synchronize]

jobs:
  # ...
```

We might want to always trigger builds for the master branch, or for every tag that is created. But for pull requests
(which have their own webhook, and therefore, trigger), we can specify that only creation or updates to those Pull
Requests' history should trigger the job.

## Manual workflows with comments

We can also react to comments on the Pull Request itself, which can be useful if you want a more seamless integration of
some features.

In my case, I wanted to deploy a preview version of Pull Requests to our frontend application.  It wasn't efficient to
do this for every single PR though (only a small subset of them actually need this), so we went with this instead:

> ![Comments triggering GitHub Action](./github-actions-comment.png)

Reviewer 2: Wait, what causes the trigger? Is it the 🚀 emoji? Is it a simple comment. Or the "preview" keyword?

Author: uh. seems I accidentally deleted one sentence here. I distinctly remember writing it

Will add back. but FYI: it's the "preview" keyword. the emoji is the action answering back to acknowledge it got the message
